webapp/maps_heatmaps/src/clickInfoWindow.ts

```diff
@@ -0,0 +1,178 @@
+      dataref[i].data().labels,
+      year,
+      month
+    );
+  }
+
+  //TODO: after runing mocha tests on the browser, remove this function out of setFirstTwentyMarkers.
+  /* placing a marker with a click event in a given location. 
+   when clicking on the marker an infoWindow will appear 
+   with all the information on this location from the database. */
+  function addMarkerWithListener(
+    map: google.maps.Map,
+    latlng: google.maps.LatLng,
+    labels: string[],
+    year?: number,
+    month?: number
+  ) {
+    const InfoWindow: google.maps.InfoWindow = new google.maps.InfoWindow();
+    const marker = new google.maps.Marker({
+      position: latlng,
+      map: map,
+    });
+    markers.push(marker);
+    google.maps.event.addListener(marker, "click", async () => {
+      const center = convertLatlngToGeopoint(marker.getPosition());
+      if (center !== undefined) {
+        const dataref = await (
+          await getQueriedCollection(center, 0, labels, year, month).get()
+        ).docs[0];
+        InfoWindow.setContent(
+          getInfoWindowContent(
+            getLabels(dataref.data()),
+            getUrl(dataref.data()),
+            getDate(dataref.data()),
+            getAttribution(dataref.data())
+          )
+        );
+        if (lastInfoWindow !== null) lastInfoWindow.close();
+        InfoWindow.open(map, marker);
+        lastInfoWindow = InfoWindow;
+      }
+    });
+  }
+}
+
+/* this function renders information for the info window, in html format.
+   @param mylabels - all the labels this image has, 
+   @param myurl - the image url,
+   @param mydate - the date the image was taken,
+   @param myattribution - who uplouded the image, */
+function getInfoWindowContent(
+  mylabels: string,
+  myurl: string,
+  mydate: string,
+  myattribution: string
+): string {
+  const contentString =
+    "<div class='map-info-window'>" +
+    "<img class='fit-image' src=" +
+    myurl +
+    "></div>" +
+    "<div style='float:right; padding: 10px;'>" +
+    "<p class='fit-text'>" +
+    mylabels +
+    mydate +
+    myattribution +
+    "</p></div>";
+  return contentString;
+}
+
+function getLabels(dataref: GeoFirestoreTypes.GeoDocumentData) {
+  let content = "<b>Labels: </b>";
+  dataref.labels.forEach((element: string) => {
+    content += element;
+    content += " ";
+  });
+  content += "<br/>";
+  return content;
+}
+
+function getDate(dataref: GeoFirestoreTypes.GeoDocumentData) {
+  const content =
+    "<b>Date: </b>" +
+    dataref.day +
+    "/" +
+    dataref.month +
+    "/" +
+    dataref.year +
+    "<br/>";
+  return content;
+}
+
+function getUrl(dataref: GeoFirestoreTypes.GeoDocumentData) {
+  const content = dataref.url;
+  return content;
+}
+
+//TODO: add field attribute to database and extract attribute field in this function.
+function getAttribution(dataref: GeoFirestoreTypes.GeoDocumentData) {
+  const content = "<b>Attribution: </b>uploader of the image_______ <br/>";
```

Comment:
make a React Component for this

Reply:
I used react.
A function in InfoWindowContent react component:)